File: api/scripts/seed_demo_content/distributions.go
// Distribution helpers for realistic engagement simulation.
//
// All randomness flows through an explicit *rand.Rand so the caller
// controls determinism. Seed once with rand.New(rand.NewSource(42))
// in main and re-runs produce identical activity data — same as the
// Faker.seed(42) convention in the Phase 3a Python distributions.
package main

import (
	"math"
	"math/rand"
	"time"
)

// Academic-calendar month weights (sum = 12.00). Same numbers as the
// Phase 3a Python `DEFAULT_MONTH_WEIGHTS` so re-seeding either layer
// produces the same time-distribution shape.
//
// Pattern: low in Jun-Jul (summer), steady during semester weeks,
// peak in May + December (exam weeks).
var defaultMonthWeights = [12]float64{
	1.10, // Jan — post-break ramp-up
	1.00, // Feb
	1.00, // Mar
	1.00, // Apr
	1.50, // May — spring exam peak
	0.50, // Jun — summer low
	0.45, // Jul — summer low
	1.20, // Aug — fall ramp-up
	1.00, // Sep
	1.00, // Oct
	1.10, // Nov
	1.65, // Dec — fall exam peak
}

// zipfDistribution returns n integers summing to ~total following Zipf's law.
//
// Exponent s controls skew: 1.0 classic, 1.2 moderate (top 5 of 82
// hold ~55% with floor=5), 2.0 severe. Each value >= floor.
//
// Pure function — deterministic given inputs. Caller decides whether
// to shuffle before assigning to entities (typical: shuffle so the
// "popular" buckets don't correlate with insert order).
func zipfDistribution(n int, s float64, total int, floor int) []int {
	if n <= 0 {
		return nil
	}

	raw := make([]float64, n)
	var norm float64
	for i := range n {
		raw[i] = 1.0 / math.Pow(float64(i+1), s)
		norm += raw[i]
	}

	counts := make([]int, n)
	rest := total - floor*n
	for i := range n {
		c := int(math.Round(float64(rest)*raw[i]/norm)) + floor
		if c < floor {
			c = floor
		}
		counts[i] = c
	}

	// Adjust the head so sum exactly matches total — rounding errors
	// accumulate on n items; fix on the biggest bucket where ±1 is invisible.
	var sum int
	for _, c := range counts {
		sum += c
	}
	if drift := total - sum; drift != 0 {
		counts[0] += drift
	}
	return counts
}

// longTailCount samples a positive integer from an exponential distribution
// with mean ~mean. Use for per-entity activity counts where most items get
// roughly the mean and a handful go viral.
//
// floor guarantees a minimum (prevents "guide has 0 votes" empty states).
// cap=0 means uncapped.
func longTailCount(rng *rand.Rand, mean float64, floor int, cap int) int {
	if mean <= 0 {
		return floor
	}
	sample := int(rng.ExpFloat64() * mean)
	if sample < floor {
		sample = floor
	}
	if cap > 0 && sample > cap {
		sample = cap
	}
	return sample
}

// backdatedTimestamp samples a UTC timestamp in [start, end] weighted
// by month-of-year per defaultMonthWeights.
//
// Two-step: pick a month-window weighted by its weight, then pick a
// uniform offset within that month's intersection with [start, end].
func backdatedTimestamp(rng *rand.Rand, start, end time.Time) time.Time {
	type window struct {
		lo, hi time.Time
		weight float64
	}
	var windows []window
	cur := time.Date(start.Year(), start.Month(), 1, 0, 0, 0, 0, time.UTC)
	for !cur.After(end) {
		next := cur.AddDate(0, 1, 0)
		lo := cur
		if start.After(lo) {
			lo = start
		}
		hi := next
		if end.Before(hi) {
			hi = end
		}
		if lo.Before(hi) {
			windows = append(windows, window{lo: lo, hi: hi, weight: defaultMonthWeights[cur.Month()-1]})
		}
		cur = next
	}
	if len(windows) == 0 {
		return uniformInWindow(rng, start, end)
	}

	var total float64
	for _, w := range windows {
		total += w.weight
	}
	if total == 0 {
		return uniformInWindow(rng, start, end)
	}
	pick := rng.Float64() * total
	var acc float64
	for _, w := range windows {
		acc += w.weight
		if pick <= acc {
			return uniformInWindow(rng, w.lo, w.hi)
		}
	}
	last := windows[len(windows)-1]
	return uniformInWindow(rng, last.lo, last.hi)
}

func uniformInWindow(rng *rand.Rand, lo, hi time.Time) time.Time {
	span := hi.Sub(lo).Seconds()
	offset := rng.Float64() * span
	return lo.Add(time.Duration(offset * float64(time.Second)))
}

// voteSplit allocates `total` votes into (up, down) with the rest dropped
// (cancel votes). Defaults: 80% up, 15% down, 5% cancel.
func voteSplit(rng *rand.Rand, total int, upShare, downShare float64) (up, down int) {
	if total <= 0 {
		return 0, 0
	}
	up = int(float64(total) * upShare)
	down = int(float64(total) * downShare)
	for range total - up - down {
		pick := rng.Float64()
		switch {
		case pick < 0.05:
			// cancel
		case pick < 0.05+upShare:
			up++
		default:
			down++
		}
	}
	return up, down
}

File: api/scripts/seed_demo/seed_demo/seeder/distributions.py
"""Distribution helpers for producing realistic engagement data.

Every function that uses randomness takes an explicit `random.Random`
instance (typically `Faker.random`). This keeps determinism at the call
site — the caller controls the seed, the module just does the math.

Four families of helpers:

1. `zipf_distribution`        — engagement counts (views, vote totals)
2. `long_tail_count`          — per-entity activity counts
3. `backdated_timestamp`      — academic-calendar-weighted timestamps
4. `vote_split`               — allocate a vote budget into up/down/cancel

The academic-calendar weights in `DEFAULT_MONTH_WEIGHTS` are opinionated:
December + May peak (exam weeks), August-November + February-April are
steady, June-July are low (summer). Override via the `weights` kwarg if
a non-academic product uses these helpers later.
"""

from __future__ import annotations

import random
from dataclasses import dataclass
from datetime import UTC, datetime, timedelta

# ---------------------------------------------------------------------------
# Academic-calendar month weights (sum = 12.00)
# ---------------------------------------------------------------------------

# Relative weights per calendar month. Pattern: low in June-July (summer),
# steady during semester weeks, peak in May + December (exam weeks).
DEFAULT_MONTH_WEIGHTS: tuple[float, ...] = (
    1.10,  # Jan — post-break ramp-up
    1.00,  # Feb
    1.00,  # Mar
    1.00,  # Apr
    1.50,  # May — spring exam peak
    0.50,  # Jun — summer low
    0.45,  # Jul — summer low
    1.20,  # Aug — fall ramp-up
    1.00,  # Sep
    1.00,  # Oct
    1.10,  # Nov
    1.65,  # Dec — fall exam peak
)


# ---------------------------------------------------------------------------
# Zipf (engagement counts)
# ---------------------------------------------------------------------------


def zipf_distribution(n: int, s: float, total: int, floor: int = 0) -> list[int]:
    """Return n integers summing to ~total, following Zipf's law.

    Exponent `s` controls skew:
      - 1.0 → classic Zipf (long tail)
      - 1.2 → moderately skewed (top 5 of 82 hold ~55% of mass with floor=5)
      - 2.0 → severely skewed (top item dominates)

    Each returned value is >= floor.

    Pure function — deterministic given inputs. No RNG; the caller
    decides whether to `rng.shuffle()` the result to decorrelate from
    insert order.
    """
    if n <= 0:
        return []
    if total < floor * n:
        raise ValueError(f"total ({total}) must be >= floor ({floor}) * n ({n}) = {floor * n}")

    raw = [1.0 / ((i + 1) ** s) for i in range(n)]
    norm = sum(raw)
    # Proportional allocation with floor enforced.
    counts = [max(floor, round((total - floor * n) * r / norm) + floor) for r in raw]

    # Adjust the head to match `total` exactly — rounding errors accumulate;
    # fix them on the biggest bucket where a ±1 is invisible.
    drift = total - sum(counts)
    if drift != 0:
        counts[0] += drift

    return counts


# ---------------------------------------------------------------------------
# Long-tail per-entity counts (votes, sessions)
# ---------------------------------------------------------------------------


def long_tail_count(
    rng: random.Random, *, mean: float, floor: int = 0, cap: int | None = None
) -> int:
    """Sample a positive integer from an exponential long-tail distribution.

    Produces a mean of ~`mean` (with an exponential right tail). Use for
    things like "how many votes does this guide have" where most items
    get the mean-ish and a handful go viral.

    `floor` guarantees a minimum (prevents "this guide has 0 of X" empty
    states). `cap` truncates runaway samples if set.
    """
    if mean <= 0:
        raise ValueError(f"mean must be positive, got {mean}")
    # expovariate(lambda) has mean = 1/lambda
    sample = int(rng.expovariate(1.0 / mean))
    sample = max(floor, sample)
    if cap is not None:
        sample = min(cap, sample)
    return sample


# ---------------------------------------------------------------------------
# Backdated timestamps
# ---------------------------------------------------------------------------


@dataclass(frozen=True)
class TimeWindow:
    """A time window for backdating. Both endpoints are inclusive."""

    start: datetime
    end: datetime

    def __post_init__(self) -> None:
        if self.start >= self.end:
            raise ValueError(f"start ({self.start}) must be before end ({self.end})")


def window_ending_now(*, months_back: int = 12, now: datetime | None = None) -> TimeWindow:
    """Convenience: a `months_back`-wide window ending at `now` (default: UTC now)."""
    end = now if now is not None else datetime.now(tz=UTC)
    # Approximate month arithmetic — good enough for demo data (30.44 days avg).
    start = end - timedelta(days=int(months_back * 30.44))
    return TimeWindow(start=start, end=end)


def backdated_timestamp(
    rng: random.Random,
    *,
    window: TimeWindow,
    month_weights: tuple[float, ...] = DEFAULT_MONTH_WEIGHTS,
) -> datetime:
    """Sample a timestamp from `window`, weighted by month-of-year.

    Two-step sampling: pick a month (weighted by month_weights), then
    pick a uniform offset within that month's intersection with `window`.

    Falls back to uniform-over-window if no month intersects (e.g.
    window is narrower than a month, or weights are all zero).
    """
    if len(month_weights) != 12:
        raise ValueError(f"month_weights must have 12 entries, got {len(month_weights)}")

    # Enumerate every (month, year) pair that overlaps the window + compute
    # the overlap interval. Weight each pair by month_weights[month-1].
    overlaps: list[tuple[datetime, datetime, float]] = []
    cursor = datetime(window.start.year, window.start.month, 1, tzinfo=window.start.tzinfo)
    while cursor <= window.end:
        if cursor.month == 12:
            next_cursor = cursor.replace(year=cursor.year + 1, month=1)
        else:
            next_cursor = cursor.replace(month=cursor.month + 1)
        lo = max(cursor, window.start)
        hi = min(next_cursor, window.end)
        if lo < hi:
            weight = month_weights[cursor.month - 1]
            overlaps.append((lo, hi, weight))
        cursor = next_cursor

    if not overlaps or sum(w for _, _, w in overlaps) == 0:
        return _uniform_in(rng, window.start, window.end)

    weights = [w for _, _, w in overlaps]
    lo, hi, _ = rng.choices(overlaps, weights=weights, k=1)[0]
    return _uniform_in(rng, lo, hi)


def _uniform_in(rng: random.Random, lo: datetime, hi: datetime) -> datetime:
    span = (hi - lo).total_seconds()
    offset = rng.uniform(0, span)
    return lo + timedelta(seconds=offset)


# ---------------------------------------------------------------------------
# Vote splitting (for study_guide_votes allocation)
# ---------------------------------------------------------------------------


def vote_split(
    rng: random.Random,
    total: int,
    *,
    up_share: float = 0.80,
    down_share: float = 0.15,
) -> tuple[int, int]:
    """Split a total vote budget into (up, down) counts.

    Remaining share (~5% by default) is dropped — simulates votes that
    the user cast then cancelled. Caller discards that portion.

    Rounding rule: `up` takes the floor of its share; `down` takes the
    rounded share; the leftover (if any) is allocated to up/down with
    additional rng draws so the sum exactly matches up+down allocated.
    """
    if total < 0:
        raise ValueError(f"total must be non-negative, got {total}")
    if up_share + down_share > 1.0 + 1e-9:
        raise ValueError(f"up_share + down_share must be <= 1.0, got {up_share + down_share}")

    up_count = int(total * up_share)
    down_count = int(total * down_share)
    leftover = total - up_count - down_count
    # Small stochastic correction to avoid everything rounding the same way.
    for _ in range(leftover):
        pick = rng.random()
        if pick < 0.05:
            continue  # cancel
        elif pick < 0.05 + up_share:
            up_count += 1
        else:
            down_count += 1
    return up_count, down_count
